File: src/main/com/lupus/opener/gui/selectables/SelectableTop.java
package com.lupus.opener.gui.selectables;

import com.lupus.command.LupusCommandFrameWork;
import com.lupus.gui.SelectableItem;
import com.lupus.gui.utils.ItemUtility;
import com.lupus.gui.utils.SkullUtility;
import com.lupus.gui.utils.TextUtility;
import com.lupus.opener.chests.MinecraftCase;
import com.lupus.opener.gui.top.GUITopCase;
import com.lupus.opener.messages.Message;
import com.lupus.opener.messages.MessageReplaceQuery;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class SelectableTop extends SelectableItem {
	UUID player;
	MinecraftCase minecraftCase;
	public SelectableTop(ItemStack item, MinecraftCase minecraftCase, UUID player) {
		super(true, item);
		this.player = player;
		this.minecraftCase = minecraftCase;
	}

	@Override
	protected void execute(Object... objects) {
		Player p = Bukkit.getPlayer(player);
		if (p == null){
			return;
		}
		List<Map.Entry<UUID,Integer>> keys =  minecraftCase.getTopKeys();
		if (keys == null){
			p.sendMessage(Message.CASE_IS_CALCULATING_TOP.toString());
			return;
		}
		if (keys.size() <= 0){
			p.sendMessage(Message.CASE_HAS_0_KEYS.toString());
			return;
		}
		int limit = Math.min(keys.size(), 50)+1;
		ItemStack[] items = new ItemStack[limit];
		for (int i = 1; i < limit; i++) {
			var entry = keys.get(i-1);

			OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(entry.getKey());
			String nickName = offlinePlayer.getName();

			if (nickName == null)
				nickName = Message.UNKNOWN_NICKNAME.toString();
			var mrq = new MessageReplaceQuery().
					addQuery("place",i+"").
					addQuery("name",nickName);
			nickName = Message.PLAYER_PLACE_TEMPLATE.toString(mrq);
			ItemStack skull = SkullUtility.getSkullFromPlayer(entry.getKey());
			ItemUtility.setItemTitle(skull,nickName);
			mrq = new MessageReplaceQuery().
					addQuery("amount",entry.getValue() + "");
			ItemUtility.setItemLore(skull,new String[]{Message.PLAYER_CASE_KEY_FORMAT.toString(mrq)});
			if(skull == null)
				continue;

			items[i-1] = skull;
		}
		ItemStack filler = new ItemStack(Material.BLUE_STAINED_GLASS_PANE);
		String[] lore = Message.SELECTABLE_TOP_FILLER_LORE.toString().split("\\n");
		ItemUtility.setItemTitleAndLore(filler,Message.SELECTABLE_TOP_FILLER_NAME.toString(), Arrays.asList(lore));

		var gui = new GUITopCase(TextUtility.color(minecraftCase.getOfficialName()),limit,filler,items);
		gui.open(p);
	}
}

File: src/main/com/lupus/opener/gui/selectables/SelectableTop.kt
package com.lupus.opener.gui.selectables

import com.lupus.gui.TopPyramidGUI
import com.lupus.opener.chests.MinecraftCase
import java.util.UUID
import com.lupus.gui.SelectableItem
import com.lupus.gui.IGUI
import com.lupus.gui.utils.SkullUtility
import com.lupus.gui.utils.ItemUtility
import java.util.Arrays
import com.lupus.opener.gui.top.GUITopCase
import com.lupus.gui.utils.TextUtility
import com.lupus.gui.PlayerSelectableItem
import com.lupus.opener.gui.ItemEditor
import com.lupus.gui.Paginator
import java.text.DecimalFormat
import com.lupus.opener.managers.ChestManager
import com.lupus.opener.gui.selectables.SelectableCase
import net.luckperms.api.LuckPerms
import com.lupus.opener.CaseOpener
import net.luckperms.api.query.QueryOptions
import net.luckperms.api.query.QueryMode
import com.lupus.opener.gui.BuyCaseGUI
import com.lupus.opener.gui.selectables.SelectableCommand
import com.lupus.gui.GUI
import com.lupus.opener.gui.BuyKeysCMD
import com.lupus.gui.utils.InventoryUtility
import com.lupus.opener.chests.CaseItem
import com.lupus.opener.gui.selectables.SelectableTop
import com.lupus.opener.gui.TopKeysGUI
import com.lupus.opener.gui.OpeningCase
import com.lupus.opener.managers.OpenerManager
import com.lupus.opener.chests.CaseItemHolder
import com.lupus.opener.gui.selectables.SelectableItemEditor
import com.lupus.opener.chests.utils.MinecraftCaseUtils
import com.lupus.gui.utils.NBTUtility
import java.util.HashMap
import java.util.TreeMap
import com.lupus.opener.chests.PlayerKey
import com.lupus.opener.gui.CaseItemList
import com.lupus.opener.runnables.ChestOpener
import java.lang.StringBuilder
import java.util.LinkedList
import java.lang.Runnable
import com.lupus.command.framework.commands.PlayerCommand
import com.lupus.opener.commands.sub.admin.GetCaseCMD
import kotlin.Throws
import com.lupus.command.framework.commands.LupusCommand
import com.lupus.command.framework.commands.CommandMeta
import com.lupus.opener.commands.sub.admin.GiveKeyCMD
import com.lupus.opener.commands.sub.admin.SetIconCMD
import com.lupus.opener.commands.sub.admin.OpenCaseCMD
import com.lupus.opener.commands.sub.admin.ReloadAllCMD
import com.lupus.opener.commands.sub.admin.RemoveKeyCMD
import com.lupus.opener.commands.sub.admin.SaveCasesCMD
import com.lupus.opener.commands.sub.admin.EditWeightCMD
import com.lupus.opener.commands.sub.admin.GetCobblexCMD
import com.lupus.opener.commands.sub.admin.OpenEditorCMD
import com.lupus.opener.gui.ChestList
import com.lupus.opener.commands.sub.admin.ResetAccountCMD
import com.lupus.opener.commands.sub.admin.CreateNewCaseCMD
import com.lupus.opener.commands.sub.admin.AllowDestructionCMD
import com.lupus.opener.listeners.BlockManipulationListener
import com.lupus.opener.commands.sub.admin.SetStatTrackCommand
import com.lupus.opener.commands.sub.player.KeysCMD
import com.lupus.opener.chests.MinecraftKey
import com.lupus.opener.commands.sub.player.BuyKeyCMD
import com.lupus.opener.commands.sub.player.KeyTopCMD
import com.lupus.opener.commands.sub.player.ChangeKeyCMD
import com.lupus.opener.commands.sub.player.RandomCaseDaily
import java.time.Instant
import com.lupus.opener.commands.sub.player.GetCraftedCobblex
import com.lupus.opener.commands.sub.player.KeyTransactionCMD
import com.lupus.command.framework.commands.arguments.UInteger
import com.lupus.opener.commands.sub.player.WithdrawKeyCommand
import java.lang.IllegalArgumentException
import java.util.HashSet
import com.lupus.command.framework.commands.SupCommand
import com.lupus.command.framework.commands.PlayerSupCommand
import com.lupus.opener.commands.PlayerCaseCommand
import com.lupus.opener.runnables.ChestSave
import com.lupus.gui.utils.ConfigUtility
import org.bukkit.plugin.java.annotation.plugin.author.Author
import org.bukkit.plugin.java.annotation.plugin.Website
import org.bukkit.plugin.java.annotation.plugin.ApiVersion
import com.lupus.opener.listeners.PvEListener
import com.lupus.opener.listeners.InventoryListener
import com.lupus.command.framework.commands.arguments.ArgumentRunner
import com.lupus.opener.messages.Message
import com.lupus.opener.messages.MessageReplaceQuery
import net.milkbowl.vault.economy.Economy
import net.luckperms.api.LuckPermsProvider
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

class SelectableTop(item: ItemStack?, var minecraftCase: MinecraftCase, var player: UUID?) :
    SelectableItem(true, item) {
    var gui: IGUI? = null
    override fun execute(vararg objects: Any) {
        val p = objects[0] as Player ?: return
        if (gui == null) loadIn()
        gui!!.open(p)
    }

    private fun loadIn() {
        val keys = minecraftCase.topKeys
        val limit = Math.min(keys!!.size, 50) + 1
        val items = arrayOfNulls<ItemStack>(limit)
        for (i in 1 until limit) {
            val entry = keys[i - 1]
            val offlinePlayer: OfflinePlayer = Bukkit.getOfflinePlayer(entry.key ?: UUID.randomUUID())
            var nickName = offlinePlayer.name
            if (nickName == null) nickName = Message.UNKNOWN_NICKNAME.toString()
            var mrq = MessageReplaceQuery().addQuery("place", i.toString() + "").addQuery("name", nickName)
            nickName = Message.PLAYER_PLACE_TEMPLATE.toString(mrq)
            val skull = SkullUtility.getSkullFromPlayer(entry.key)
            ItemUtility.setItemTitle(skull, nickName)
            mrq = MessageReplaceQuery().addQuery("amount", entry.value.toString() + "")
            ItemUtility.setItemLore(skull, arrayOf<String>(Message.PLAYER_CASE_KEY_FORMAT.toString(mrq)))
            if (skull == null) continue
            items[i - 1] = skull
        }
        val filler = ItemStack(Material.BLUE_STAINED_GLASS_PANE)
        val lore = Message.SELECTABLE_TOP_FILLER_LORE.toString().split("\\n".toRegex()).toTypedArray()
        ItemUtility.setItemTitleAndLore(filler, Message.SELECTABLE_TOP_FILLER_NAME.toString(), Arrays.asList(*lore))
        gui = GUITopCase(TextUtility.color(minecraftCase.officialName1), limit, filler, *items)
    }

    init {
        loadIn()
    }
}